### PR TITLE
BaseService: Query params

### PR DESCRIPTION
**Pull request by paulmelnikow**
This is a fairly straightforward improvement to `BaseService` which provides support for query parameters, in parity with the functionality in old-style services

This functionality is necessary to port many of the existing services, and arose while I was working on this WIP branch: https://github.com/badges/shields/compare/master...paulmelnikow:rewrite-npm-typedefs

### CODE REVIEW

**Comment by chris48s** [comment]
@platan do you mind taking the review of this one?

**Comment by paulmelnikow** [comment]
It would be really great if our [bus factor](https://en.wikipedia.org/wiki/Bus_factor) on this BaseService code were greater than 1. It's so essential to how Shields works! If the code is hard to understand, it's important we make it clearer… and all the more reason to address those issues before the problem becomes even harder to tackle.

@platan Can I implore your review? 👐

**Comment by platan** [comment]
@paulmelnikow I'm starting to read it ...

File: services/base.spec.js [hunk 1]
```diff
@@ -9,18 +9,17 @@ const BaseService = require('./base');
...
+    return { message: `Hello ${someArg}${suffix}` };
   }
 
   static get category() { return 'cat'; }
   static get url() {
     return {
       base: 'foo',
       format: '([^/]+)',
-      capture: ['someArg']
+      capture: ['someArg'],
```
Comment (platan) [inline services/base.spec.js]: Can we rename `someArg` to `namedArgA` (or `namedArg1`).

File: services/base.spec.js [hunk 2]
```diff
@@ -9,18 +9,17 @@ const BaseService = require('./base');
...
+      queryParams: ['suffix'],
```
Comment (platan) [inline services/base.spec.js]: Can we rename `suffix` to `queryParamA`?

File: services/base.spec.js [hunk 3]
```diff
@@ -9,18 +9,17 @@ const BaseService = require('./base');
 require('../lib/register-chai-plugins.spec');
 
 class DummyService extends BaseService {
-  async handle({someArg}) {
-    return {
-      message: 'Hello ' + someArg,
-    };
+  async handle({ someArg }, { suffix }) {
+    return { message: `Hello ${someArg}${suffix}` };
```
Comment (platan) [inline services/base.spec.js]: How about this?
```js
return { message: `Hello namedParamA${namedParamA} with queryParamA${quaryParamA}` };
```

File: services/base.spec.js [hunk 4]
```diff
@@ -145,14 +146,14 @@ describe('BaseService', () => {
...
-      const requestHandler = mockHandleRequest.getCall(0).args[0];
+      const { handler: requestHandler } = mockHandleRequest.getCall(0).args[0];
 
       const mockSendBadge = sinon.spy();
       const mockRequest = {
         asPromise: sinon.spy(),
       };
       await requestHandler(
-        /*data*/ {},
+        /*queryParams*/ { suffix: '?' },
```
Comment (platan) [inline services/base.spec.js]: What do you think about extracting the first and the second arguments to variables?
```js
      const queryParams = { suffix: '?' };
      const match = '/foo/bar.svg'.match(expectedRouteRegex);
      await requestHandler(
        queryParams,
        match,
        mockSendBadge,
        mockRequest
      );
```

File: services/base.spec.js [hunk 5]
```diff
@@ -162,7 +163,7 @@ describe('BaseService', () => {
       expect(mockSendBadge).to.have.been.calledWith(
         /*format*/ 'svg',
```
Comment (platan) [inline services/base.spec.js]: How  about extracting `svg` to a variable?

File: services/base.js [hunk 6]
```diff
@@ -38,6 +38,9 @@ module.exports = class BaseService {
    *  - capture: Array of names for the capture groups in the regular
    *             expression. The handler will be passed an object containing
    *             the matches.
+   *  - queryParams: Array of names for query parameters which will the service
```
[the remark this answers is not in the document]
Comment (paulmelnikow) [reply services/base.js]: Thanks, looks like this was addressed in #1582.

File: services/base.spec.js (hunk 3 above)
Comment (paulmelnikow) [reply services/base.spec.js]: Works for me!

**Comment by paulmelnikow** [comment]
Thank you so much for the review!

> But changes in test code were difficult to understand for me. `DummyService` does nothing specific and you have to memorize it internals before analyzing test code. I've left some suggestions, they could improve understanding of the code. 

I'm totally open to improving this. Maybe we could think about creating a dummy service which has clearer semantics.

**Pull request merged**